Add support for inlining new atomic methods on Power

hzongaro opened the pull request:
Added support for inlining of the new atomic intrinsics
atomic{AddAndFetch|FetchAndAdd|Swap}{32|64}BitSymbol on Power.  This is
largely a copy of similar support in OpenJ9 for inlining methods of
AtomicInteger and AtomicLong on Power.

Issue:  #2958

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

hzongaro commented:
May I ask @0dvictor @zl-wang @gita-omr to review?

One thing I'll point out is that I put `OMR::Power::CodeGenerator::inlineDirectCall` in `OMRTreeEvaluator.cpp` simply because the corresponding `J9::Power::CodeGenerator::inlineDirectCall` is defined in `J9TreeEvaluator.cpp`, rather than in `J9CodeGenerator.cpp`, which feels odd.

There will be a separate PR in OpenJ9 to take advantage of this change.

0dvictor commented on compiler/p/codegen/OMRCodeGenerator.hpp:
```diff
@@ -171,6 +171,8 @@ class OMR_EXTENSIBLE CodeGenerator : public OMR::CodeGenerator
 
    bool getSupportsIbyteswap();
 
+   bool supportsAtomicAdd() {return true;}
```
FYI, https://github.com/eclipse/omr/pull/2999 is changing this query, it may need to be updated before merging.

hzongaro replied:
Revised in commit https://github.com/eclipse/omr/pull/3018/commits/8077bb18b5170cb1e62f23fef75258c0dc006d5d

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   if (deltaChild->getOpCode().isLoadConst()
+          && !deltaChild->getRegister()
+          && deltaChild->getDataType() == TR::Int32)
```
Could the second argument be 64bit constant and within [UPPER, LOWER] range at the same time? Don't want to miss that opt opportunity.

hzongaro replied:
Sounds good.  Thanks, Julian!

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+   conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions((uint16_t) numDeps, (uint16_t) numDeps, cg->trMemory());
+
+   addDependency(conditions, valueAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   conditions->getPreConditions()->getRegisterDependency(0)->setExcludeGPR0();
```
No need to exclude gr0 from this register, since you are explicitly putting it in the indexReg position.

hzongaro replied:
Thanks, Julian @zl-wang .  Just to clarify, are you suggesting I remove just line 5308, or both lines 5308 and 5309?  Sorry for the basic questions - I'm not really familiar yet with how register dependencies are specified or how they work.

zl-wang replied:
@hzongaro you can eliminate both 5308 and 5309.

hzongaro replied:
Addressed in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+      else
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, deltaReg, resultReg, deltaReg);
+      }
+
```
Maybe I misunderstood it, but AddAndFetch result should be the sum?  Here, you have the loaded value as the result always.

hzongaro replied:
@zl-wang Sorry for the confusion, Julian.  At lines #5338-5352, the code will return `resultReg` for `FetchAndAdd` and `deltaReg` for `AddAndFetch` - that's based on the value of `isGetThenUpdate`.  Would it help if I changed the name of `resultReg` to `prevValueReg`, say, and renamed `deltaReg` to `updatedValueReg`?

hzongaro replied:
Fixed up variable names to clarify what they hold.  Revised in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+      // For non-constant arguments, use evaluated register
+      // For non-immediate constants, evaluate since they may be re-used
+      numDeps++;
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, deltaReg, tempReg);
```
Should explicitly test if it will be re-used ... and you potentially can save this copy instruction.

hzongaro replied:
@zl-wang Julian, I assume you mean only for the non-immediate constant case.  I think you're suggesting that in the case of a non-immediate constant, it should evaluate the constant in the loop, but for a non-constant, it should be evaluated outside of the loop, the value copied into another register to using an `mr`.  Do I have that right?

Now that I'm thinking about it, why does this code try to reuse the register that holds delta to hold the sum. . . ?

zl-wang replied:
@hzongaro  certainly the new sequence is better.  previously, my comment, looking at the mr instruction, meant to say why not testing the delta child referenceCount to decide if delta can be reused.  In your new sequence, if  deltaChild->getReferenceCount() == 1, you can even eliminate the creation of GPR_88 and use GPR_99 in its place.  But this is not strictly necessary.

hzongaro replied:
Revised generated code so that the `mr` in the loop is no longer used.  Changed in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+         // avoid evaluating immediates for add operations
+         isDeltaImmediate = true;
+         }
+      else if (delta & 0xFFFF == 0 && (delta & 0xFFFF0000) >> 16 <= UPPER_IMMED && (delta & 0xFFFF0000) >> 16 >= LOWER_IMMED)
```
For a signed 32bit value, you can test  delta & 0xffff == 0  and don't need to test the latter two clauses (they are true always).

hzongaro replied:
Eliminated the redundant test.  Fixed in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   return doInline;
+   }
+
```
High-level issue:  this will not work for 32bit JVM.  i.e.  this is assuming a 64bit JVM.

hzongaro replied:
@zl-wang Julian, thanks for catching this.  Just so I'm clear, can this code be inlined on a 32 bit JVM when it's working with 32 bit operands, and inlined on a 64 bit JVM when it's working with either 32 bit and 64 bit operands?  Or can it only be inlined on a 64 JVM ever?

Something I just noticed this morning, I was playing with `AtomicLong.getAndAdd` and `AtomicInteger.getAndAdd` without my changes for these new intrinsics, and it looks like inline code is never generated for `AtomicLong.getAndAdd` for 32 bit or 64 bit JVMs, and always generated for `AtomicInteger.getAndAdd` for both 32 bit and 64 bit JVMs.  Is that expected?

zl-wang replied:
@hzongaro I didn't expect that.  I expected existing JVM can inline CAS of both 32/64 operands in both 32/64 bit JVM.  The only assumption is the underlying CPU is 64bit (this has been true since java7.1 when we discontinued support of 32bit hardware).  The case for inlining CAS of 64bit operand in 32bit JVM needs special care though ... you can assemble 64bit operand in a register within the CAS range only, depending on the fact that CAS will definitely fail if there was a signal delivered in the between (i.e.  meaning the upper 32bit of the register can be trashed anytime).

hzongaro replied:
@zl-wang Julian, I've been working on a revision that handles 64-bit values running in a 32-bit JVM, but it makes the code quite ugly.  I’ll pull that out as a separate pull request, and I’ll let you decide whether it’s an important enough case to handle, as well as point out any problems with the way that I tackled the problem.

hzongaro commented:
@zl-wang @fjeremic Julian, Filip - I've made revisions that I think addresses your review comments.  May I ask you to re-review when you have some time?  The one exception is handling 64-bit values on a 32-bit JVM, which I will submit in a follow up PR.

fjeremic commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+      loadConstant(cg, node, delta, deltaReg);
+      }
+
+   uint8_t len = isLong ? 8 : 4;
```
This should be `const`

hzongaro replied:
Corrected in commit https://github.com/eclipse/omr/pull/3018/commits/cf67e3ee2bc9ec34b0170de4e4dc302555d71b9a

hzongaro commented on compiler/p/codegen/OMRCodeGenerator.hpp:
```diff
@@ -384,6 +386,9 @@ class OMR_EXTENSIBLE CodeGenerator : public OMR::CodeGenerator
    bool supportsNegativeFusedMultiplyAdd() {return true;}
 
+   bool supportsNonHelper(TR::SymbolReferenceTable::CommonNonhelperSymbol symbol
+);
```
Oops!  Didn't notice that working with an 80 column window.

hzongaro replied:
Corrected in commit https://github.com/eclipse/omr/pull/3018/commits/cf67e3ee2bc9ec34b0170de4e4dc302555d71b9a

fjeremic commented:
@hzongaro I think you accidentally checked in some huge file as part of your latest commit.

zl-wang commented on compiler/p/code:
```diff
@@ -0,0 +1,3682 @@
+bool OMR::Power::CodeGenerator::supportsAESInstructions()
+   {
+    if ( TR::Compiler->target.cpu.getPPCSupportsAES() && !self()->comp()->getOption(TR_DisableAESInHardware))
+      ret
```
This looks like you accidentally copied OMRCodeGenerator.cpp to this compiler/p/code file?

zl-wang commented on compiler/p/codegen/OMRCodeGenerator.cpp:
```diff
@@ -3660,3 +3660,23 @@ bool OMR::Power::CodeGenerator::getSupportsIbyteswap()
+      case TR::SymbolReferenceTable::atomicAddSymbol:
+      case TR::SymbolReferenceTable::atomicFetchAndAddSymbol:
+      case TR::SymbolReferenceTable::atomicSwapSymbol:
+         {
```
Does it matter that you missed cases here of data-type-specific symbols, e.g., atomicSwap32BitSymbol?

hzongaro replied:
Thanks for raising that question.  I based this on the implementations of `OMR::X86::CodeGenerator::supportsNonHelper` and `OMR::Z::CodeGenerator::supportsNonHelper`.  @0dvictor @fjeremic Victor, Filip - should `supportsNonHelper` include the 32-bit and 64-bit-specific nonhelper symbols?

0dvictor replied:
I believe the data-type-specific ones are deprecated and will be deleted soon. AFAIK, none of them is in-use as of now.

hzongaro replied:
Thank you!

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+   // Memory barrier --- NOTE: we should be able to do a test upfront to save this barrier,
+   //                          but Hursley advised to be conservative due to lack of specification.
+   generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
```
I am wondering these symbols are only intended to map to the corresponding Unsafe methods.  If yes, the note on memory barriers is justified.  If not, the note is pessimistic regarding memory barriers.

0dvictor replied:
They are not. Unsafe is only one of many use cases of these symbols; other use cases include automatic reference count in Swift, as well as similar synchronization helpers for other languages.

hzongaro replied:
I spoke with Julian, and he indicated that his comment was about the use of memory barriers themselves rather than about the comment in the code.  He indicated that as these symbols are used for `Unsafe` (in addition to other uses) the memory barriers are required.

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+          && !deltaChild->getRegister()
+          && (deltaChild->getDataType() == TR::Int32
+                 || (deltaChild->getDataType() == TR::Int64
+                        && deltaChild->getInt() == deltaChild->getLongInt())))
```
This test is most likely incorrect to one of the BigEndian or LittleEndian cases, i.e.  it cannot be correct for both BE and LE.

hzongaro replied:
Fixed in https://github.com/eclipse/omr/pull/3018/commits/a453a47d5b03cd028f0c9be8e0ed55b752ea2d70

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+   generateDepLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel, conditions);
+
+   cg->decReferenceCount(valueAddrChild);
+   cg->stopUsingRegister(valueAddrReg);
```
As a general rule, you don't need to stop-using a register if it is carried in a node, since decReferenceCount on the node automatically takes care of the liveness.  Otherwise, if the node is referenced by multiple nodes, you run the risk of prematurely ending its life.

fjeremic replied:
I think an assert would have caught this in DEBUG mode. Another reason why we need to start moving towards running testing with asserts enabled like we used to in the past.

hzongaro replied:
Thanks, Julian!  That really clarifies for me the proper use of those methods.  Fixed in https://github.com/eclipse/omr/pull/3018/commits/a453a47d5b03cd028f0c9be8e0ed55b752ea2d70

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   if (deltaReg)
+      {
+      cg->stopUsingRegister(deltaReg);
```
Same issue here as above.

hzongaro replied:
Fixed in https://github.com/eclipse/omr/pull/3018/commits/a453a47d5b03cd028f0c9be8e0ed55b752ea2d70

zl-wang commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   if (deltaChild)
+      {
+      cg->decReferenceCount(deltaChild);
```
Furthermore, deltaReg might be returned as resultReg.  You had better setRegister before this decReferenceCount, because it can prematurely end that register's life (if this were the last reference to deltaChild).

hzongaro replied:
Fixed in https://github.com/eclipse/omr/pull/3018/commits/a453a47d5b03cd028f0c9be8e0ed55b752ea2d70

hzongaro commented:
I believe all review comments have been addressed, so I will remove "WIP" from the title.

fjeremic commented:
@genie-omr build plinux,aix
